File: ts/src/resources/marketplace.ts
import { HttpTransport } from "../http.js"

export interface EmailTemplate {
  _id: string
  name?: string
  subject?: string
  body?: string
  [key: string]: unknown
}

export interface CreateEmailTemplateInput {
  name: string
  subject?: string
  body?: string
  [key: string]: unknown
}

export interface ChannelWorkflowInput {
  channel_id?: string
  workflow_id?: string
  [key: string]: unknown
}

export interface ChannelWorkflowResponse {
  success: boolean
  [key: string]: unknown
}

export interface InstallFromJsonInput {
  template?: Record<string, unknown>
  [key: string]: unknown
}

export interface InstallFromJsonResponse {
  success: boolean
  [key: string]: unknown
}

export interface MarketplaceFileDetails {
  _id: string
  name?: string
  url?: string
  [key: string]: unknown
}

export interface MarketplaceFileUploadResponse {
  url: string
  file_id?: string
  [key: string]: unknown
}

export class MarketplaceResource {
  /**
   * @param base    - Marketplace service base URL (gateway/marketplaces/v2)
   * @param gateway - Gateway root URL (kept for legacy download fallback)
   */
  constructor(
    private readonly http: HttpTransport,
    private readonly base: string,
    private readonly gateway: string,
  ) {}

  private get root() { return this.base.replace(/\/$/, "") }

  // ── Templates (marketplace use-case templates) ─────────────────────────

  async listUseCaseTemplates(): Promise<{ _id: string; name?: string; [key: string]: unknown }[]> {
    return this.http.getFetch()(`${this.root}/market-places/v2/templates`, { method: "GET" }).then(r => r.json())
  }

  async installFromJson(body: InstallFromJsonInput): Promise<InstallFromJsonResponse> {
    return this.http.getFetch()(`${this.root}/market-places/templates/install-from-json`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ── Files (marketplace-scoped file uploads / downloads) ────────────────

  async uploadFile(body: FormData): Promise<MarketplaceFileUploadResponse> {
    return this.http.getFetch()(`${this.root}/files`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  async deleteFile(fileId: string): Promise<void> {
    await this.http.getFetch()(`${this.root}/files/${fileId}`, { method: "DELETE" })
  }

  async getFileDetails(fileId: string): Promise<MarketplaceFileDetails> {
    return this.http.getFetch()(`${this.root}/file-details/${fileId}`, { method: "GET" }).then(r => r.json())
  }

  async downloadMarketPlaceFile(shortPath: string): Promise<Response> {
    return this.http.getFetch()(`${this.root}/files/${shortPath}`, { method: "GET" })
  }

  // ── Email templates ────────────────────────────────────────────────────

  async listEmailTemplates(params?: Record<string, string>): Promise<EmailTemplate[]> {
    const url = new URL(`${this.root}/email-templates/search`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async createEmailTemplate(body: CreateEmailTemplateInput): Promise<EmailTemplate> {
    return this.http.getFetch()(`${this.root}/email-templates`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ── Channel workflows ──────────────────────────────────────────────────

  async postChannelWorkflows(body: ChannelWorkflowInput): Promise<ChannelWorkflowResponse> {
    return this.http.getFetch()(`${this.root}/market-places/channel-workflows`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }
}

File: py/src/imbrace/resources/marketplace.py
from typing import Any, Dict, Optional
from ..http import HttpTransport, AsyncHttpTransport


class MarketplaceResource:
    """Marketplace domain — Sync.

    Base URL is the marketplace service (`{gateway}/marketplaces/v2`).
    Routes hit the marketplace microservice's own router (NOT legacy backend).
    """

    def __init__(self, http: HttpTransport, base: str, gateway: Optional[str] = None):
        self._http = http
        self._root = base.rstrip("/")
        # `gateway` kept for back-compat; unused after v1.1.0.
        self._gateway = gateway.rstrip("/") if gateway else None

    # --- Templates ---
    def list_use_case_templates(self) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._root}/market-places/v2/templates").json()

    def install_from_json(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._root}/market-places/templates/install-from-json", json=body).json()

    # --- Files ---
    def upload_file(self, files: Any) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._root}/files", files=files).json()

    def delete_file(self, file_id: str) -> None:
        self._http.request("DELETE", f"{self._root}/files/{file_id}")

    def get_file_details(self, file_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._root}/file-details/{file_id}").json()

    def download_market_place_file(self, short_path: str) -> Any:
        return self._http.request("GET", f"{self._root}/files/{short_path}")

    # --- Email Templates ---
    def list_email_templates(self, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._root}/email-templates/search", params=params or {}).json()

    def create_email_template(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._root}/email-templates", json=body).json()

    # --- Channel workflows ---
    def post_channel_workflows(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._root}/market-places/channel-workflows", json=body).json()


class AsyncMarketplaceResource:
    """Marketplace domain — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str, gateway: Optional[str] = None):
        self._http = http
        self._root = base.rstrip("/")
        self._gateway = gateway.rstrip("/") if gateway else None

    async def list_use_case_templates(self) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._root}/market-places/v2/templates")
        return res.json()

    async def install_from_json(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._root}/market-places/templates/install-from-json", json=body)
        return res.json()

    async def upload_file(self, files: Any) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._root}/files", files=files)
        return res.json()

    async def delete_file(self, file_id: str) -> None:
        await self._http.request("DELETE", f"{self._root}/files/{file_id}")

    async def get_file_details(self, file_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._root}/file-details/{file_id}")
        return res.json()

    async def download_market_place_file(self, short_path: str) -> Any:
        return await self._http.request("GET", f"{self._root}/files/{short_path}")

    async def list_email_templates(self, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._root}/email-templates/search", params=params or {})
        return res.json()

    async def create_email_template(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._root}/email-templates", json=body)
        return res.json()

    async def post_channel_workflows(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._root}/market-places/channel-workflows", json=body)
        return res.json()
